lib/models/application.js

```diff
@@ -0,0 +1,1618 @@
+/*
+Copyright 2016 Balena
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+   http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+import * as url from 'url';
+
+import * as Promise from 'bluebird';
+const once = require('lodash/once');
+import * as errors from 'balena-errors';
+
+import {
+	isId,
+	isNoApplicationForKeyResponse,
+	isNotFoundResponse,
+	findCallback,
+	mergePineOptions,
+	treatAsMissingApplication,
+	LOCKED_STATUS_CODE,
+} from '../util';
+
+import { normalizeDeviceOsVersion } from '../util/device-os-version';
+import {
+	getCurrentServiceDetailsPineOptions,
+	generateCurrentServiceDetails,
+} from '../util/device-service-details';
+
+const getApplicationModel = function(deps, opts) {
+	const { request, pine } = deps;
+	let { apiUrl, dashboardUrl } = opts;
+
+	const deviceModel = once(() => require('./device').default(deps, opts));
+	const releaseModel = once(() => require('./release').default(deps, opts));
+
+	const { buildDependentResource } = require('../util/dependent-resource');
+
+	const tagsModel = buildDependentResource(
+		{ pine },
+		{
+			resourceName: 'application_tag',
+			resourceKeyField: 'tag_key',
+			parentResourceName: 'application',
+			getResourceId(nameOrSlugOrId) {
+				return exports.get(nameOrSlugOrId, { $select: 'id' }).get('id');
+			},
+		},
```

Comment:
I had to remove unused params from these, see https://github.com/balena-io/balena-sdk/pull/867